### PR TITLE
Generate containers unit test source files

### PR DESCRIPTION
**Pull request by dalg24**

### CODE REVIEW

**Comment by dalg24** [comment]
Make sure we update the Makefile too https://github.com/kokkos/kokkos/blob/b46f8b22c0b276718a14a58bb6f825ed748166a9/containers/unit_tests/Makefile

File: containers/unit_tests/CMakeLists.txt [hunk 1]
```diff
@@ -13,23 +13,29 @@ foreach(Tag Threads;Serial;OpenMP;HPX;Cuda)
...
-        ${dir}/Test${Tag}_OffsetView.cpp
-        ${dir}/Test${Tag}_ScatterView.cpp
-        ${dir}/Test${Tag}_StaticCrsGraph.cpp
-        ${dir}/Test${Tag}_UnorderedMap.cpp
-        ${dir}/Test${Tag}_Vector.cpp
-        ${dir}/Test${Tag}_ViewCtorPropEmbeddedDim.cpp
-      )
+    set(UnitTestSources UnitTestMain.cpp)
+    set(dir ${CMAKE_CURRENT_BINARY_DIR}/${dir})
+    file(MAKE_DIRECTORY ${dir})
```
Comment (jjwilke) [inline containers/unit_tests/CMakeLists.txt:18]: Was there a reason for making a directory rather prepending dir_ to the filename? Doesn't need to change. Just curious.

File: containers/unit_tests/CMakeLists.txt [hunk 2]
```diff
@@ -13,23 +13,29 @@ foreach(Tag Threads;Serial;OpenMP;HPX;Cuda)
...
+        ErrorReporter
+        OffsetView
+        ScatterView
+        StaticCrsGraph
+        UnorderedMap
+        Vector
+        ViewCtorPropEmbeddedDim
+        )
+      set(file ${dir}/Test${Tag}_${Name}.cpp)
+      file(WRITE ${file} "#include <Test${Tag}_Category.hpp>\n")
```
Comment (jjwilke) [inline containers/unit_tests/CMakeLists.txt:35]: Probably never going to really matter, but better to limit the number of file operations:
````
      string(CONCAT file_text
        "#include <Test${Tag}_Category.hpp>"
        "\n#include <Test${Name}.hpp>"
      )
      file(WRITE ${file} ${file_text})
````

File: containers/unit_tests/Makefile [hunk 3]
```diff
@@ -31,14 +31,24 @@ KOKKOS_CXXFLAGS += -I$(GTEST_PATH) -I${KOKKOS_PATH}/containers/unit_tests
...
 ifeq ($(KOKKOS_INTERNAL_USE_CUDA), 1)
 	OBJ_CUDA = UnitTestMain.o gtest-all.o
-	OBJ_CUDA += TestCuda_BitSet.o
+	OBJ_CUDA += TestCuda_Bitset.o
 	OBJ_CUDA += TestCuda_DualView.o
 	OBJ_CUDA += TestCuda_DynamicView.o
-	OBJ_CUDA += TestCuda_DynRankViewAPI_generic.o
-	OBJ_CUDA += TestCuda_DynRankViewAPI_rank12345.o
-	OBJ_CUDA += TestCuda_DynRankViewAPI_rank67.o
+	OBJ_CUDA += TestCuda_DynViewAPI_generic.o
```
Comment (jjwilke) [inline containers/unit_tests/Makefile:49]: I don't know enough bash foo to see how to fix it, but something just feels wrong when we generate source files in a nice, clean loop above... then explicitly add each object file here.

File: containers/unit_tests/cuda/TestCuda_BitSet.cpp [hunk 4]
```diff
@@ -1,47 +0,0 @@
-
-/*
-//@HEADER
-// ************************************************************************
-//
```
Comment (jjwilke) [inline containers/unit_tests/cuda/TestCuda_BitSet.cpp:5]: Just for fun, I'd love to know what Sandia legal would say about license headers in generated code.

File: containers/unit_tests/CMakeLists.txt (hunk 1 above)
Comment (dalg24) [reply containers/unit_tests/CMakeLists.txt:18]: No reason

File: containers/unit_tests/CMakeLists.txt (hunk 2 above)
Comment (dalg24) [reply containers/unit_tests/CMakeLists.txt:35]: Good point will change.

File: containers/unit_tests/Makefile (hunk 3 above)
Comment (crtrott) [reply containers/unit_tests/Makefile:49]: I can't figure it out. Inside a foreach loop I can apparently not append variables.

**Pull request merged**